lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp

```diff
@@ -0,0 +1,330 @@
+  // Ignore connections between different types.
+  if (connect.src().getType() != connect.dest().getType())
+    return false;
+  // Ignore if either source or destination is an argument.
+  if (!connect.dest().getDefiningOp() || !connect.src().getDefiningOp())
+    return false;
+  return true;
+}
+
+// Return the constant value if value is essentially constant.
+static Value getConstant(Value value) {
+  if (auto constant = value.getDefiningOp<ConstantOp>())
+    return constant;
+  if (auto constant = value.getDefiningOp<InvalidValueOp>())
+    return constant;
+  if (auto bitcast = value.getDefiningOp<BitCastOp>())
+    return getConstant(bitcast.input());
+
+  // TODO: Add unrealized_conversion, asUInt, asSInt
+  return {};
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+// A helper struct to merge connections.
+struct MergeConnection {
+
+  // This function peels a pair of src and dest which might be possible to
+  // merge. More specifically, this function recognizes these patterns and
+  // push %a and %b to the worklist.
+  //    connect: subfield %a <= subfield %b
+  //    connect: subindex %a <= subindex %b
+  bool peelFullyConnect(ConnectOp connect);
+
+  // This function peels a destination which might be possible to
+  // merge to constant. More specifically, this function recognizes these
+  // patterns and push %a to the worklist.
+  //    connect: subfield %a <= constant
+  //    connect: subindex %a <= constant
+  bool peelConstantConnect(ConnectOp connect);
+
+  template <class AggregateType>
+  void tryFullyConnect(AggregateType type, Value src, Value dest);
+
+  template <class AggregateType>
+  void tryConstantConnect(AggregateType type, Value dest);
+
+  // Return true if a connection is removed.
+  bool run(FModuleOp moduleOp);
+
+  // A set to track investigated destinations.
+  DenseSet<FieldRef> visitedDestination;
+
+  // A map from a destination FieldRef to connect op.
+  DenseMap<FieldRef, ConnectOp> fieldRefToConnect;
+
+  // A map from a destination FieldRef to fieldRef of its connected source
+  // value.
+  DenseMap<FieldRef, FieldRef> connectedSource;
+
+  // A map from a destination FieldRef to Constant-like value for its
+  // connecetd source value.
+  DenseMap<FieldRef, Value> connectedConstant;
+
+  // A work list for fully connection merging. We use a queue to traverse leafs
+  // first.
+  std::queue<std::pair<Value, Value>> fullyConnectWorklist;
+
+  // A work list for constant connection merging. We use a queue to traverse
+  // leafs first.
+  std::queue<Value> constantConnectWorklist;
+
+  // Container for unused connections to erase in the post-process.
+  SmallVector<ConnectOp> shouldBeRemoved;
+};
+
+// Peel if the connect might be a part of full connections.
+bool MergeConnection::peelFullyConnect(ConnectOp connect) {
```

Comment:
Would it be advantageous to restructure this as in the first pass collecting all connects by source (with map source -> set of connects), then testing each element in the map to see if it has a complete set of connects?

Reply:
Thanks, I restructured in a similar way.